File: src/test/java/com/leetcode/p0651_four_keys_keyboard/SolutionTest.java
package com.leetcode.p0651_four_keys_keyboard;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

class SolutionTest {

    static Stream<Solution> solutions() {
        return Stream.of(
                new Default()
        );
    }

    // --- Provided examples ---

    @ParameterizedTest
    @MethodSource("solutions")
    void example1_nEquals3(Solution solution) {
        // A, A, A => 3
        assertEquals(3, solution.maxA(3));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void example2_nEquals7(Solution solution) {
        // A, A, A, Ctrl-A, Ctrl-C, Ctrl-V, Ctrl-V => 9
        assertEquals(9, solution.maxA(7));
    }

    // --- Small values ---

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals1_singleKeystroke(Solution solution) {
        assertEquals(1, solution.maxA(1));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals2(Solution solution) {
        assertEquals(2, solution.maxA(2));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals4(Solution solution) {
        // A, A, A, A => 4 (no benefit from copy-paste yet)
        assertEquals(4, solution.maxA(4));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals5(Solution solution) {
        // A, A, A, A, A => 5 (copy-paste doesn't help at n=5)
        assertEquals(5, solution.maxA(5));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals6(Solution solution) {
        // A, A, A, Ctrl-A, Ctrl-C, Ctrl-V => 6
        assertEquals(6, solution.maxA(6));
    }

    // --- Medium values ---

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals8(Solution solution) {
        assertEquals(12, solution.maxA(8));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals9(Solution solution) {
        assertEquals(16, solution.maxA(9));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals10(Solution solution) {
        assertEquals(20, solution.maxA(10));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals11(Solution solution) {
        assertEquals(27, solution.maxA(11));
    }

    // --- Larger values ---

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals15(Solution solution) {
        assertEquals(81, solution.maxA(15));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals20(Solution solution) {
        assertEquals(324, solution.maxA(20));
    }

    // --- Upper bound ---

    @ParameterizedTest
    @MethodSource("solutions")
    void nEquals50_upperBound(Solution solution) {
        assertEquals(1327104, solution.maxA(50));
    }
}

File: problems/p0651_four_keys_keyboard/typescript/_solution.test.ts
import { describe, test, expect } from "vitest";
import { solutions } from "./solution";

describe.each(solutions)("maxA - %o", (maxA) => {
  // --- Provided examples ---

  test("example 1: n=3 => 3 (A,A,A)", () => {
    expect(maxA(3)).toBe(3);
  });

  test("example 2: n=7 => 9 (A,A,A,Ctrl-A,Ctrl-C,Ctrl-V,Ctrl-V)", () => {
    expect(maxA(7)).toBe(9);
  });

  // --- Small values ---

  test("n=1, single keystroke", () => {
    expect(maxA(1)).toBe(1);
  });

  test("n=2", () => {
    expect(maxA(2)).toBe(2);
  });

  test("n=4, no benefit from copy-paste yet", () => {
    expect(maxA(4)).toBe(4);
  });

  test("n=5", () => {
    expect(maxA(5)).toBe(5);
  });

  test("n=6, first time copy-paste ties with just typing", () => {
    expect(maxA(6)).toBe(6);
  });

  // --- Medium values ---

  test("n=8", () => {
    expect(maxA(8)).toBe(12);
  });

  test("n=9", () => {
    expect(maxA(9)).toBe(16);
  });

  test("n=10", () => {
    expect(maxA(10)).toBe(20);
  });

  test("n=11", () => {
    expect(maxA(11)).toBe(27);
  });

  // --- Larger values ---

  test("n=15", () => {
    expect(maxA(15)).toBe(81);
  });

  test("n=20", () => {
    expect(maxA(20)).toBe(324);
  });

  // --- Upper bound ---

  test("n=50, upper constraint", () => {
    expect(maxA(50)).toBe(1327104);
  });
});
